src/pydicom/pixels/decoders/base.py

```diff
@@ -0,0 +1,1911 @@
+            * (frames, rows, columns) for multi-frame, single plane data
+            * (frames, rows, columns, planes) for multi-frame, multi-plane data
+
+            A writeable :class:`~numpy.ndarray` is returned by default. For
+            native transfer syntaxes with ``view_only=True`` a read-only
+            :class:`~numpy.ndarray` will be returned if `src` is immutable.
+        """
+        if not HAVE_NP:
+            raise ImportError(
+                "NumPy is required when converting pixel data to an ndarray"
+            )
+
+        if index is not None and index < 0:
+            raise ValueError("'index' must be greater than or equal to 0")
+
+        runner = DecodeRunner(self.UID)
+        runner.set_source(src)
+        runner.set_options(**kwargs)
+        runner.set_decoders(self._validate_decoders(decoding_plugin))
+
+        if config.debugging:
+            LOGGER.debug(runner)
+
+        if validate:
+            runner.validate_options()
+            runner.validate_buffer()
+
+        if self.is_native:
+            func = self._as_array_native
+            as_writeable = not runner.get_option("view_only", False)
+        else:
+            func = self._as_array_encapsulated
+            as_writeable = True
+
+        arr = runner.reshape(
+            func(runner, index),
+            as_frame=False if index is None else True,
+        )
+
+        if raw:
+            return arr.copy() if not arr.flags.writeable and as_writeable else arr
+
+        # Processing may give us a new writeable array anyway, so do
+        #   it first to avoid an unnecessary ndarray.copy()
+        arr = runner.process(arr)
+
+        return arr.copy() if not arr.flags.writeable and as_writeable else arr
+
+    @staticmethod
+    def _as_array_encapsulated(runner: DecodeRunner, index: int | None) -> "np.ndarray":
+        """Return compressed and encapsulated pixel data as :class:`~numpy.ndarray`.
+
+        Parameters
+        ----------
+        runner : pydicom.pixels.decoders.base.DecodeRunner
+            The runner with the encoded data and decoding options.
+        index : int | None
+            The index of the frame to be returned, or ``None`` if all frames
+            are to be returned
+
+        Returns
+        -------
+        numpy.ndarray
+            A 1D array containing the pixel data.
+        """
+        dtype = runner.pixel_dtype
+
+        # Return the specified frame only
+        if index is not None:
+            return np.frombuffer(runner.decode(index=index), dtype=dtype)
+
+        # Return all frames
+        # Preallocate container array for the frames
+        pixels_per_frame = runner.frame_length(unit="pixels")
+        arr = np.empty(pixels_per_frame * runner.number_of_frames, dtype=dtype)
+        for idx, frame in enumerate(runner.iter_decode()):
+            start = idx * pixels_per_frame
+            arr[start : start + pixels_per_frame] = np.frombuffer(frame, dtype=dtype)
```

Comment:
Consider checking if more frames remain, and concatenating to the pre-allocated array.

It's more expensive, but probably pretty rare.